### DISCUSSION TITLE
How to do a color reduction of a scanned image into 3 precise colors?

### DISCUSSION THREAD

**Discussion by smaekelen** [ImageMagick/ImageMagick#4342] [Help]
I would like to clean a scanned image like the one displayed here below. I want to convert the red text/graphics to pure red color and the black to pure black, without any color gradients. Ultimately, I am looking for a color reduction method that reduces the scanned image to 3 colors(red, black and white). Obviously I want the text and the graphics to have smooth edges too. Thank you in advance for any suggestions.
![Image0003](https://user-images.githubusercontent.com/91072416/136628341-f49ae792-44c6-4b78-8622-d84caaf6d4fc.jpg)

**Comment by fmw42** [comment]
One cannot ensure smooth edges and only 3 colors. You are essentially asking to remove the anti-aliasings. So you will end up with aliased graphics.

What version of ImageMagick? What platform/OS?  Syntax may differ. 

Here is for ImageMagick 6 on Unix-like systems.

`convert img.jpg \( xc:white xc:red xc:black +append +write mpr:ctable +delete \) -dither none -remap mpr:ctable result.gif
`

![result](https://user-images.githubusercontent.com/13458218/136630101-3641b825-48b6-4d2e-882f-5c95e9215ad8.gif)

**Comment by smaekelen** [comment]
Thank you @fmw42 for your suggestion.
I am using Imagemagick 7.1, in Manjaro/Arch-Linux. I guess what I need is anti-aliasing with smooth edges. I would like to make a pdf from these cleaned images at the end.
I just tried something:
```
mogrify -path out2 -modulate 100,2000 *.tif
mogrify -path out3 -enhance -despeckle -fill black -opaque yellow -fuzz 40% *.tif
```
The result is not perfect, but similar to what I would like to have. The image has some noise. Is there something better that I can do?

![Image0003-2](https://user-images.githubusercontent.com/91072416/136653056-84874536-066c-428c-8ac4-170890b244b4.jpg)

**Reply by fmw42** [reply]
This image is a JPG not tif and has a file size 1.4 MB.  With Imagemagick 7, you should be using magick mogrify or magick if replacing convert. Please post an example TIF input and explain further what you are trying to achieve and what output format you are trying to have. It is not clear to me whether you want fewer colors or anti-aliasing.

**Reply by smaekelen** [reply]
This is what I am trying to achieve: I scanned a book and I would like to make a high quality pdf from that, without having a big file size. The book uses only red, black and white colors. I thought of doing a color reduction using imagemagick, reducing the colors used in the images to these 3 colors. In this way I hope to reduce the size of the pdf file, while maintaining good image quality. May be I am going the wrong route to do the whole thing. I have used scantailor to preprocess the scanned images. Here are 2 example TIF files:
https://i7.putstuffonline.com/02m25FCmzec/putstuffonline.tif, https://i7.putstuffonline.com/9XCk2Z8JJcc/putstuffonline.tif ( Github does not permit me to upload tif files, that is why I uploaded JPG in my earlier posts.)
I would like to get a tif output format for the processed images, so that I can run an OCR and then make a pdf out of it.
@fmw42, I just saw that you have a website with a lot of imagemagick scripts. One of them is for processing text files. Will that be of any use in my case?

**Reply by fmw42** [reply]
You can zip the tif file and Git will allow that to be uploaded. Which of my script are you referring? I doubt they do what you want.  Have you considered raster to vector conversion?  Getting small size and limited colors are competing things. They generally do not go together, since anti-aliasing produces new colors.

**Reply by smaekelen** [reply]
OK. Thanks for the tip to upload the file.
I was referring to your TEXTCLEANER script. 
I have tried some raster to vector conversion in the past (some years ago), but I never had a good quality vector output on text images. If you have some suggestions I don't mind trying.

**Reply by fmw42** [reply]
textcleaner is useful to whiten background when the background is dark and varying.

**Reply by smaekelen** [reply]
Which software would you suggest for vectorization?

**Comment by snibgo** [comment]
The red ink is semi-transparent, so where the ink is thicker, we get a darker red.

> I would like to get a tif output format for the processed images, so that I can run an OCR and then make a pdf out of it.

Is OCR the objective? What OCR product are you using? I get a fairly good result from Tesseract, if I darken the red ink first:
```
magick putstuffonline2.tif -colorspace Gray -level 70,100% g.tiff

tesseract g.tiff g
```
The output g.txt starts with:
```
lxiv

Ritus servandus in celebratione Missaa

disjunguntur. In omni etiam deosculatio-
ne sive Altaris, sive libri, sive alterius rei,
non producitur signum Crucis pollice
vel manu super id quod osculandum est.

2. Osculato Altari, a_ccedit ad cornu
ej us sinistrum, id est Epistolae: ubi stans
versus Altare et producens a fronte ad
pectus signum â€˜Crucis, incipit intelli-
gibili voce Introitum Missae et prosequi-
tur junctis manibus. Cum dicit: Gloria
Patri, tenens junctas manus, caput incli-
nat versus Crucem. Cum repetit Introi-
tum, non signat se, ut prius, et eo
repetito, junctis manibus ante pectus
accedit ad medium Altaris, ubi stans
versus illud similiter manibus junctis,
dicit eadem voce ter Kvrie, elÃ©ison, ter
Christe, elÃ©ison, et iterum ter K)"rie,
elÃ©ison alternatim cum ministro. Si mi-
nister, vel qui intersunt Celebranti non
respondeant, ipse solus novies dicit.
```
If your PDF is for reading by humans, then remember that humans are good at ignoring the actual colour of "white" paper, ignoring ink splashes, filling gaps in letters, and so on.

**Reply by smaekelen** [reply]
Thank you @snibgo for your suggestions. I am sorry for the lack of clarity. OCR is not the main objective, but to have a good quality pdf file that can be also used for a reprint and of smaller file size.
I have AbbyyFinereader to OCR.
